### DISCUSSION TITLE
megaTinyCore's UPDI uploader is a lot slower than pyupdi

### DISCUSSION THREAD

**Discussion by lerouxb** [SpenceKonde/megaTinyCore#358] [Show and tell]
I did some testing using arduino-cli (because I just can't even with that IDE - trying ways to use my own editor):

```
lerouxb@updawg:~/Documents/Arduino/DoubleBufferedMillis$ time arduino-cli upload -b "megaTinyCore:megaavr:atxy6:bodmode=disabled,bodvoltage=1v8,chip=3216,clock=20internal,eesave=enable,millis=enabled,resetpin=UPDI,serialevent=no,startuptime=8,uartvoltage=5v" -P pyupdi -p /dev/cu.usbserial-DA01L36P
Pinging device...
Ping response: 1E9521
Setting fuse 0x2=0x2
Writing literal values...
Verifying literal values...
Setting fuse 0x6=0x4
Writing literal values...
Verifying literal values...
Setting fuse 0x8=0x0
Writing literal values...
Verifying literal values...
Finished writing fuses.
Chip/Bulk erase,
Memory type eeprom is conditionally erased (depending upon EESAVE fuse setting)
Memory type flash is always erased
Memory type internal_sram is always erased
Memory type lockbits is always erased
...
Erased.
Writing from hex file...
Writing flash...
[==================================================] 39/39
Verifying flash...
OK
Verifying...
Verify successful. Data in flash matches data in specified hex-file

real	0m52.262s
user	0m0.714s
sys	0m0.196s
```

vs

```
lerouxb@updawg:~/Documents/Arduino/DoubleBufferedMillis/build/megaTinyCore.megaavr.atxy6$ time pyupdi -d tiny3216 -c /dev/cu.usbserial-DA01L36P -f DoubleBufferedMillis.ino.hex
Device info: {'family': 'tinyAVR', 'nvm': 'P:0', 'ocd': 'D:0', 'osc': '3', 'device_id': '1E9521', 'device_rev': '0.2'}
Programming successful

real	0m13.900s
user	0m0.261s
sys	0m0.087s
```

I compiled like this:

```
lerouxb@updawg:~/Documents/Arduino/DoubleBufferedMillis$ arduino-cli compile -b "megaTinyCore:megaavr:atxy6:bodmode=disabled,bodvoltage=1v8,chip=3216,clock=20internal,eesave=enable,millis=enabled,resetpin=UPDI,serialevent=no,startuptime=8,uartvoltage=5v" -e
Sketch uses 4966 bytes (15%) of program storage space. Maximum is 32768 bytes.
Global variables use 99 bytes (4%) of dynamic memory, leaving 1949 bytes for local variables. Maximum is 2048 bytes.
```

Required a bit of guessing to get that `fqbn` string. Ended up fishing the board config options/values out of my arduino preferences. I _think_ it is correct, but I plan to do the ifdef and error checks as outlined in the readme just in case.

Just an FYI. Maybe documentation that might help others that want to try and do this.

**Comment by SpenceKonde** [comment]
Okay - well, we considered using pyupdi, but we were warned that that would be the Wrong Way - two microchip engineers, and the author of pyupdi were unanimous in their opinion that it should be based on pymcuprog instead - the general impression they gave is that pyupdi was not production grade and was designed more as a demo than something that should be a real programming tool - while pymcuprog, since it gets official quality control from Microchip, was a far more robust solution. We all know how rigorous Microchip quality control is, right? Anyway, we weren't about to disregard their words, and did as they advised

If you want to write up documentation on how to use it with the cli - awesome I would love to include it. However I do not have time to do that write up. My to-do list is a mile long, and I've got enough work for a month that needs to be done yesterday if not sooner.
